Import from Study List feature

ecxyzzy opened the pull request:
## Summary

Allows new users to AntAlmanac to import their WebReg study list, which they can then save to the backend if they so choose.

Demo (you may need to open in new tab to view it in full resolution):
![antalmanac-import-studylist](https://user-images.githubusercontent.com/89349085/158491098-6fa45e86-01f4-4457-a6fa-5526034553c8.gif)
<sub>Incidentally, if anyone knows of a dark theme for WebReg please let me know…</sub>

## Test Plan
1. Leaving the textbox blank results in an error `Snackbar` telling the user that they cannot import an empty/invalid Study List.
2. Filling the textbox with gibberish results in the same `Snackbar` as above.
3. Pasting a properly formatted Study List as copy-pasted from WebReg, with the term selector in the right-hand pane set to the corresponding term, correctly imports the Study List into the currently focused Schedule.
4. Same as 3) but with the Study List from StudentAccess, which has a drastically different format compared to the Study List from WebReg, also correctly imports the Study List into the currently focused Schedule.

~~5. Any errors due to API calls failing (e.g. due to a term mismatch causing certain section codes to not exist) will result in a `Snackbar` notifying the user of an error , and the currently focused Schedule will be cleared. Any classes that were successfully added will remain added to the currently focused Schedule (possible caveat?).~~ Deprecated as of 577f946.

5. If a section cannot be added (e.g. due to a term mismatch causing that section code to not exist), the import process will continue silently. Upon conclusion of the process, a snackbar with the appropriate variant will be displayed, along with relevant information (success => all sections imported, warning => some sections imported, error => no sections imported).
6. Typing in any number of valid section codes for the current term, as long as they are delimited by a non-numeric character, correctly imports the courses as in 3) and 4). This is not really an expected usecase, but is useful if you just have a list of course codes for whatever reason, and works properly.

## Issues
Closes #225.

## Possible Enhancements before Merging

* Theming
  * Location: ~~Is the top bar appropriate for a feature like this?~~ Still provisional but has been moved into the overflow menu on mobile as of ad5efc2.
  * Icon: ~~Are there better icons than "download?" It might be a bit confusing.~~ Changed to `PostAdd` in a208b82 with the intent to update it to `ContentPasteGo` after migrating to mui v5.
* Error Handling: ~~The import can fail because of a multitude of reasons (API failure, mismatched term, empty/invalid input, etc). Do we want to tell the user what specifically went wrong, or is it okay to just say that it failed? If it comes down to user error, do we want to further delineate what exactly they did wrong?~~ Added logging to console in 3ee4ff9.
* Clearing Schedules: ~~This might not be such a good idea if (for example) the user tries to overwrite an existing schedule with a malformed Study List import, which would cause them to lose everything in that schedule. Maybe making a copy of the existing schedule and restoring it if something goes wrong would be better?~~ Removed in 3ee4ff9.
* Term Selection: ~~Add a drop-down in the dialog so users can select the term their Study List is from.~~ Implemented in 5857ce5.
* User Experience: Add a blurb to the dialog that tells the user where to actually find their Study List, as well as what and what not to copy.

Any thoughts and/or possible additional features are greatly appreciated.

ecxyzzy committed:
implement study list importing

ecxyzzy committed:
add ImportStudyList component to top bar

ChaseC99 commented on src/components/App/CustomAppBar.js:
```diff
@@ -47,6 +48,8 @@ const CustomAppBar = (props) => {
                 <div style={{ flexGrow: '1' }} />
 
+                <ImportStudyList />
+
```
Still not sure if the AppBar is the best place for this, but for now I think it should probably be the 3rd item in the list after `Load`

ChaseC99 reviewed (commented):
This is a sweet new feature that I am 100% going to use!

I left a few comments but I'm going to defer the final review to @EricPedley since he initially created the issue.

My only other concern is that users might not know what to copy from WebReg. You have a good description for what to do, but users need a LOT of hand-holding LOL. 

Potential future follow: maybe we should make a "Tips" box above the Google Analytics privacy policy that tells users about features they might not know about (adding custom events, signing up for notifications, exporting to Google Calendar, etc., and, of course, importing from WebReg) #303

ecxyzzy committed:
move ImportStudyList component

ecxyzzy committed:
improve error handling

- Log any errors during the import process to the console
- Don't clear the focused schedule if something goes wrong

devsdevsdevs commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,122 @@
+                        .map((line) => line.match(/\d{5}/g))
+                        .filter((id) => id)
+                        .flat()
+                        .map((sectionCode) => queryWebsoc({ term: this.state.selectedTerm, sectionCodes: sectionCode }))
```
Instead of querying websoc/PP API for each section code, since a studylist will have all courses in the same term, you can build a list of section codes and submit them in a single request. Keep in mind though, this will limit you to ~8 section codes (last time that I checked), so chunk your codes into groups of 8 and submit them in more requests if you need to.

devsdevsdevs commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,122 @@
+        this.setState({ isOpen: true, selectedTerm: store.getFormData().term });
+    };
+
+    handleClose = (doImport) => {
```
Would be good to make this function async to avoid .then etc.

ecxyzzy replied:
Not exactly sure I follow—the reason I used `.then` and `.catch` on `Promise.all` was to run multiple queries in parallel, and catch any errors from them. Is there a cleaner way to write this if the function was made `async`? Even with the section code chunking you suggested (great idea by the way) there are still cases where multiple queries will have to be made.

devsdevsdevs replied:
It'd just be a simple refactor,  so you'd do await on the Promise.all instead of chaining it with .then. I think we should use just promote using async instead of promise chaining because it looks cleaner, functionally it's identical.

ecxyzzy replied:
Gotcha.

Yeah, async-await definitely looks cleaner now that I've figured out how it works. Thanks for the suggestions :D

ecxyzzy committed:
chunk codes into sets of eight and query together

ecxyzzy committed:
add term selector to ImportStudyList dialog

ecxyzzy committed:
refactor array chunking to use ternary

ecxyzzy committed:
refactor handleClose to use async-await

ecxyzzy committed:
make import errors more user friendly

- Display a success snackbar if all sections were imported
- Display a warning snackbar if some sections were imported, along with
  a reminder for the user to check the term/study list
- Display an error snackbar if no sections were imported, with a similar
  reminder

EricPedley commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,168 @@
+                    await Promise.all(
+                        sectionCodes
+                            .reduce((result, item, index) => {
+                                // WebSOC queries can have a maximum of 8 course codes in tandem
```
Where did you find this info or how did you figure this out? I tried changing the chunk size to higher numbers when testing your code and it still seems to work.

ecxyzzy replied:
devsdevsdevs mentioned it [here](https://github.com/icssc-projects/AntAlmanac/pull/302#discussion_r827536026) and it seemed to be correct based on initial testing; I did some more testing with the REST API and it appears the limit is actually 10. Thanks for pointing that out—I'll address it.

EricPedley commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,168 @@
+                    )
+                ).forEach((response) => {
+                    response.schools
+                        .map((school) => school.departments)
```
Since you aren't returning anything here, using repeated map/flatten is more confusing than just using normal loops. I recommend refactoring it for readability. See this example of going through the websoc data from `helpers.js`:
```
for (const school of jsonResp.schools) {
      for (const department of school.departments) {
          for (const course of department.courses) {
              for (const section of course.sections) {
                  addedCourses.push({
                      ...sectionCodeToInfoMapping[section.sectionCode],
                      deptCode: department.deptCode,
                      courseNumber: course.courseNumber,
                      courseTitle: course.courseTitle,
                      courseComment: course.courseComment,
                      prerequisiteLink: course.prerequisiteLink,
                      section: section,
                  });
              }
          }
      }
  }

```
Also since we're having to do this twice, it would be nice to write a function to encapsulate extracting the individual course info from a websoc query. You can do this yourself or create an issue for it if you want.

ecxyzzy replied:
Will do. Is it best practice to make a separate branch/PR for new functionality separate from this feature (e.g. this and refactoring `TermSelector` like you mentioned [here](https://github.com/icssc-projects/AntAlmanac/pull/302#discussion_r828435583)), or is it fine to just commit to this branch?

EricPedley replied:
Normally I just add it to the same branch. Not sure if that is best practice though. @ChaseC99 what are your thoughts on when to make an entirely new branch vs just lumping a change in with an existing branch that isn't entirely related?

ChaseC99 replied:
I usually try to keep my PRs as small as possible, so it's easier for others to review.

In this case, let's create a new issue, leave a todo comment in the code, and immediately open a new PR after merging this one.

That helper function will affect core functionality, so it'd probably be best to test it thoroughly in a separate PR.

ecxyzzy replied:
Done. I've also created issues #308 and #309 for this and the TermSelector issue Eric mentioned below.

EricPedley commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,168 @@
+            const currSchedule = AppStore.getCurrentScheduleIndex();
+            const sectionCodes = this.state.studyListText
+                .split('\n')
+                .map((line) => line.match(/\d{5}/g))
```
You can do this matching without splitting the string and mapping each line. Running match on the entire study list string should do the trick. See this screenshot:
![(running a regex match in the console to illustrate the point)](https://user-images.githubusercontent.com/48658337/158674972-acd3b4b5-c3e5-497e-a125-341d54ecbd51.png)
Also I think the filter is unnecessary, and also if you run the match on the entire string isntead of each line the flatten shouldn't be needed either.

EricPedley commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,168 @@
+
+    enterEvent = (event) => {
+        const charCode = event.which ? event.which : event.keyCode;
+        if (charCode === 13 || charCode === 10) {
```
Can you add a comment for what these charCodes are? I figured out that 13 is enter but in the first two references I looked at I couldn't find what charCode 10 is and had to use String.fromCharCode to figure out it was newline.

EricPedley reviewed (changes requested):
Nice job! Looks really good functionally but there are some refactors that would make the code cleaner / more readable. Feel free to push back on any of these suggestions if you think the way you did it was better (but explain why).

EricPedley commented on src/components/App/ImportStudyList.js:
```diff
@@ -0,0 +1,168 @@
+                        </DialogContentText>
+                        <InputLabel>Term</InputLabel>
+                        <Select value={this.state.selectedTerm} onChange={this.handleChange}>
+                            {termData.map((term) => (
```
This causes a console warning because each MenuItem has no unique key prop. Get the index param from map and pass it to each menu item as a key prop (See `CustomAppBar.js` lines 81-83 for a good example of this).

EricPedley replied:
The same problem also exists at TermSelector.js line 38. And now that I see that file, I'm thinking it might be appropriate to use that here instead of a new Select component? I can see that just dropping it in wouldn't work, but you could maybe modify TermSelector to take an onchange function as a prop and use that. (i.e. replace updateFormValue with the function passed as a prop).

ecxyzzy committed:
first round of refactoring

- raise section code chunking limit to 10
- remove extraneous string/array manipulation
  from matching section codes
- add comment to annotate charCodes in enterEvent

ecxyzzy committed:
second round of refactoring

- assign index from .map() as MenuItem key for Term dropdown
- add a bunch of TODOs to be addressed by future pull requests

ChaseC99 reviewed (changes requested):
Sorry for all the revision request but could you add some top/bottom margin in between term and the input field? They're a little too close right now.
<img width="574" alt="image" src="https://user-images.githubusercontent.com/29494270/158729307-3c980bb7-4f41-4dc7-882a-724d39fecadc.png">

ecxyzzy committed:
add styling to make margins less bad

ecxyzzy committed:
move import logic into callback

ecxyzzy committed:
write instructions for copypasta-ing study list

ecxyzzy committed:
remove redundant regex match by moving declaration

ecxyzzy committed:
remove extraneous event listener remover
